Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/organizer/warn/duration/TraceLocationWarnDurationViewModel.kt

```diff
@@ -0,0 +1,107 @@
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toShortTimeFormat
+import de.rki.coronawarnapp.util.TimeStamper
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.joda.time.Duration
+import org.joda.time.Instant
+import org.joda.time.LocalDate
+import org.joda.time.LocalTime
+
+class TraceLocationWarnDurationViewModel @AssistedInject constructor(
+    @Assisted private val traceLocation: TraceLocation,
+    dispatcherProvider: DispatcherProvider,
+    private val timeStamper: TimeStamper,
+) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
+
+    private val uiState = MutableStateFlow(UiState())
+    val state: LiveData<UiState> = uiState.asLiveData2()
+    val selectedDate: LocalDate get() = uiState.value.localDate
+    val selectedTime: LocalTime get() = uiState.value.localTime
+    val selectedDuration: Duration get() = uiState.value.duration
+
+    init {
+        updateUiState()
+    }
+
+    fun dateChanged(localDate: LocalDate, localTime: LocalTime) {
+        val resultTime = if (localDate == LocalDate.now() &&
+            localTime.isAfter(LocalTime.now())
+        ) LocalTime.now() else localTime
+        uiState.apply { value = value.copy(localDate = localDate, localTime = resultTime) }
+    }
+
+    fun durationChanged(duration: Duration) {
+        uiState.apply {
+            value = value.copy(
+                duration = Duration.standardMinutes(duration.standardMinutes.coerceIn(15, 1425))
+            )
+        }
+    }
+
+    private fun updateUiState() {
+        uiState.apply {
+            value = value.copy(
+                description = traceLocation.description,
+                startDateTime = traceLocation.startDate,
+                endDateTime = traceLocation.endDate
+            )
+        }
+
+        if (traceLocation.startDate != null &&
+            traceLocation.startDate != Instant.EPOCH &&
+            !traceLocation.isBeforeStartTime(timeStamper.nowUTC)
+        ) {
+            uiState.apply {
+                value = value.copy(
+                    localDate = traceLocation.startDate.toDateTime().toLocalDate(),
+                    localTime = traceLocation.startDate.toDateTime().toLocalTime()
+                )
+            }
+        }
+
+        val duration = if (traceLocation.endDate != null && traceLocation.endDate != Instant.EPOCH) {
+            getNearestFifteen(Duration(traceLocation.startDate, traceLocation.endDate).standardMinutes)
+        } else if (traceLocation.defaultCheckInLengthInMinutes != null &&
+            traceLocation.defaultCheckInLengthInMinutes > 0
+        ) {
+            getNearestFifteen(traceLocation.defaultCheckInLengthInMinutes.toLong())
+        } else {
+            Duration.standardHours(2)
+        }
+        durationChanged(duration)
```

Comment:
The default value shall be set as follows:
- if EndDate of the TraceLocation is set and is > 0, to the neartest 15 minute increment of the difference between EndDate and StartDate
- otherwise, if DefaultCheckInLengthInMinutes of the TraceLocation is set and is > 0, to the nearest 15 minute increment of DefaultCheckInLengthInMinutes
- otherwise, to 2:00